### ISSUE TITLE
Fairseq-TPU gradient accumulation does not save memory

### ISSUE THREAD

**Issue by Eric-Wallace** [pytorch/xla#1367]
## 🐛 Bug

Gradient accumulation does not save any memory on fairseq-tpu. For example, if your TPU can fit up to batch size 128, then using batch size 128 with gradient accumulation 2 will cause an OOM error.

It seems as if it uses about the same memory as if you didn't do any gradient accumulation. For example, if can fit 128 batch size, then using batch size 32 with `--update-freq=4`  works just fine. If you use `--update-freq=8` with batch size 32 then it will fail.  

To Reproduce

Run the official tutorial, https://cloud.google.com/tpu/docs/tutorials/transformer-pytorch, but set the `--update-freq=2` instead of `--updated-freq=1`.

Expected behavior

Gradient accumulation should allow you to simulate bigger batch sizes.

Environment
reproducible on XLA backend [CPU/TPU]: 16-highmem-cpu, TPUv3-8 TPU
torch_xla version: nightly or v0.5, both don't work

@taylanbil

**Comment by taylanbil** [comment]
@Eric-Wallace our barrier is hit when we yield from the loader inside `train_loop_fn`. Essentially, every operation between two barriers is appended to a graph, which is executed when we hit the barrier.

with `--update-freq=8`, we yield 8 batches from the loader, and we will then do 8 forward passes + 1 backward pass [here](https://github.com/pytorch-tpu/fairseq/blob/master/fairseq/trainer.py#L236). That graph _can_ get too big. The way I would solve this, I'd try adding a barrier (`xm.mark_step()`) at the end of this [for loop](https://github.com/pytorch-tpu/fairseq/blob/master/fairseq/trainer.py#L236) in the case where update-freq is > 1. I'll try that and see if it works when I get some time. In the meantime, if you can try and report, we would appreciate.

**Comment by Eric-Wallace** [comment]
It works, thanks. I just added

```if self.args.xla and len(samples) > 1:
       xm.mark_step()
```

to the end of that for loop.

This detail about mark_step() should probably be added to some FAQ somewhere (not sure where) for general usage, as this is different than base PyTorch.

**Comment by Eric-Wallace** [comment]
i'll leave this issue open for now

**Comment by taylanbil** [comment]
@Eric-Wallace there's some info about the barrier and more [in our api guide](https://github.com/pytorch/xla/blob/master/API_GUIDE.md)

**Comment by taylanbil** [comment]
this issue should now be resolved on the nightly. Feel free to re-open if needed.
